firedrake/cython/dmswarm.pyx

```diff
@@ -0,0 +1,505 @@
+        for i in range(nclosure):
+            entity = cell_closures[cell, i]
+            CHKERR(PetscSectionGetDof(global_numbering.sec, entity, &ndofs))
+            if ndofs > 0:
+                CHKERR(PetscSectionGetOffset(global_numbering.sec, entity, &off))
+                # The cell we're looking at the entity through is
+                # higher than the lowest cell the column touches, so
+                # we need to offset by the difference from the bottom.
+                if variable:
+                    off += offset[flat_index[k]]*(layer_extents[c, 0] - layer_extents[entity, 0])
+                for j in range(ceil_ndofs[i]):
+                    cell_nodes[cell, flat_index[k]] = off + j
+                    k += 1
+
+    CHKERR(PetscFree(ceil_ndofs))
+    CHKERR(PetscFree(flat_index))
+    return cell_nodes
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def reordered_coords(PETSc.DM swarm, PETSc.Section global_numbering, shape):
+    """Return coordinates for the swarm, reordered according to the
+    global numbering permutation for the coordinate function space.
+
+    ..note:: This is a copy and paste job from the equivalent function
+             in dmplex.pyx with the relevant modifications made for a
+             DMSwarm.
+
+    Shape is a tuple of (mesh.num_vertices(), geometric_dim)."""
+    cdef:
+        PetscInt v, vStart, vEnd, offset
+        PetscInt i, dim = shape[1]
+        np.ndarray[PetscReal, ndim=2, mode="c"] swarm_coords, coords
+
+    # get coords field - NOTE this isn't copied so make sure
+    # swarm.restoreField is called too!
+    swarm_coords = swarm.getField("DMSwarmPIC_coor").reshape(shape)
+    coords = np.empty_like(swarm_coords)
+    vStart = 0
+    vEnd = swarm.getLocalSize()
+
+    for v in range(vStart, vEnd):
+        CHKERR(PetscSectionGetOffset(global_numbering.sec, v, &offset))
+        for i in range(dim):
+            coords[offset, i] = swarm_coords[v - vStart, i]
+
+    # have to restore coords field once accessed to allow access again
+    swarm.restoreField("DMSwarmPIC_coor")
+
+    return coords
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def remove_ghosts_pic(PETSc.DM swarm, PETSc.DM plex):
+    """Remove DMSwarm PICs which are in ghost cells of a distributed
+    DMPlex.
+
+    :arg swarm: The DMSWARM which has been associated with the input
+        DMPlex `plex` using PETSc `DMSwarmSetCellDM`.
+    :arg plex: The DMPlex which is associated with the input DMSWARM
+        `swarm`
+    """
+    cdef:
+        PetscInt cStart, cEnd, ncells, i, npics
+        PETSc.SF sf
+        PetscInt nroots, nleaves
+        const PetscInt *ilocal = NULL
+        const PetscSFNode *iremote = NULL
+        np.ndarray[PetscInt, ndim=1, mode="c"] pic_cell_indices
+        np.ndarray[PetscInt, ndim=1, mode="c"] ghost_cell_indices
+
+    # check doesn't work right now
+    # assert plex is swarm.getCellDM().dm
```

Comment:
You need to check that `plex.handle == swarm.getCellDM().dm.handle` or something like that.

Reply:
It's because of a mistake in the return type of `getCellDM` in petsc4py. Been meaning to make a PR to fix

Reply:
Opened a PR. https://bitbucket.org/petsc/petsc4py/pull-requests/152/fix-incorrect-return-type-from-dmswarm

Reply:
..which is now merged.